docs/source/manual/configuration_file.rst
.. _configuration-file:

******************
Configuration file
******************

General
*******

A `yaml <https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html>`_ configuration file is required in order to run Agilepy.

It is composed by several sections and each section holds several configuration options, some of which are optional (having a default value), some others are required.

It supports environment variables (they can be used to define file system paths).

It can be created easily, calling the following static method and passing the minimal set of (required) configuration parameters.

::

    AGAnalysis.getConfiguration(
          confilepath="./agconfig.yaml", # the destination path of the configuration file
          userName="username", # the name of the flare advocate
          sourceName="OJ287", # the name of the source
          tmin=58930, # tmin
          tmax=58936, # tmax
          timetype="MJD", # time type
          glon=206.8121188769472, # glon
          glat=35.8208923457401, # glat
          outputDir="$HOME/agilepy_analysis", # the destination path of the output directory
          verboselvl=1, # the verbosity level
          evtfile="evt indexfile", # optional parameter
          logfile="log indexfile", # optional parameter
          datapath="datapath",
          userestapi=True,

    )

The method above will create the following configuration file:

.. code-block:: yaml
  
  input:
    evtfile: None
    logfile: None
    userestapi: True
    datapath: datapath
  output:
    outdir: $HOME/agilepy_analysis
    filenameprefix: analysis_product
    sourcename: sourcename
    username: username
    verboselvl: 2

  selection:  
    tmin: 54935.0
    tmax: 54936.0
    timetype: MJD
    glon: 355.447
    glat: -0.2689
    proj: ARC
    timelist: None
    irf: H0025
    filtercode: 5
    fovradmin: 0
    fovradmax: 60
    albedorad: 80
    dq: 0
    phasecode: null
    lonpole: 180
    lpointing: null
    bpointing: null
    maplistgen: "None"

  maps:
    mapsize: 40
    useEDPmatrixforEXP: false
    expstep: null
    spectralindex: 2.1
    timestep: 160
    projtype: WCS
    proj: ARC
    binsize: 0.25
    energybins:
      - 100, 10000
    fovbinnumber: 1
    offaxisangle: 30

  model:
    modelfile: null
    galmode: 1
    isomode: 1
    galcoeff: null
    isocoeff: null
    galmode2: 0
    galmode2fit: 0
    isomode2: 0
    isomode2fit: 0

  mle:
    ranal: 10
    ulcl: 2
    loccl: 95
    expratioevaluation: true
    expratio_minthr: 0
    expratio_maxthr: 15
    expratio_size: 10
    minimizertype: Minuit
    minimizeralg: Migrad
    minimizerdefstrategy: 2
    mindefaulttolerance: 0.01
    integratortype: 1
    contourpoints: 40
    edpcorrection: 0.75
    fluxcorrection: 0

  ap:
    radius: 3
    timeslot: 3600

  plotting:
    twocolumns: False

Updating the configuration options
**********************************
The user should not directly manipulate the configuration file, because the configuration file is read only once, when the 
AGBaseAnalysis constructor is called. Hence, the configuration file modification will not affect the internal configuration object.
Also, updating the values held by this object will not affect the original values written on disk.

In order to update the internal configuration object, the user can rely on the following methods:

* `getOption(optionName) <../api/analysis_api.html#api.AGBaseAnalysis-AGBaseAnalysis-getOption>`_
* `setOption(**kwargs) <../api/analysis_api.html#api.AGBaseAnalysis-AGBaseAnalysis-setOptions>`_

For example:
::

    ag.setOptions(binsize=0.50, energybins=[[100, 300], [500, 1000]])
    print(ag.getOption("energybins"))

Configuration options
*********************
This section describes the configuration options.

Section: *'input'*
==================
This section defines the input data files. The input data files are indexes: each
row holds the file system position of an actual event data/log file, together with
the time interval it refers to. If userestapi if True the selection of evtfile and logfile is not required,
Agilepy creates its own index files automatically. See more details in `this link <Ag_dataset_rest_api.html>`_.

.. csv-table::
   :header: "Option", "Description", "Type", "Required", "Default"
   :widths: 20, 100, 20, 20, 20

   evtfile, "Path to index evt file name", str, no, None
   logfile, "Path to index log file name", str, no, None
   userestapi, "If true downloads date into datapath", bool, no, "True"
   datapath, "the position of AGILE data", str, no, "None"



Section: *'output'*
===================
The output section collects options related to the output files generation and logging.

The *'outdir'* option sets the root directory of the analysis results where all output files are written.

Agilepy use two loggers, one logs messages on the console, the other writes messages on disk.
There are 4 kind of messages based on their importance factor:

  - CRITICAL: a message describing a critical problem, something unexpected, preceding a program crash or an Exception raise.
  - WARNING: an indication that something unexpected happened, or indicative of some problem in the near future (e.g. ‘disk space low’). The software is still working as expected.
  - INFO: confirmation that things are working as expected.
  - DEBUG: detailed information, typically of interest only when diagnosing problems.

The *'verboselvl'* option sets the verbosity of the Agilepy console logger. The Agilepy file logger verbosity is set to 2 by default.


.. csv-table::
   :header: "Option", "Description", "Type", "Required", "Default"
   :widths: 20, 100, 20, 20, 20

   "outdir", "Path of the output directory", "str", "yes", "null"
   "filenameprefix", "The filename prefix of each output file", "str", "yes", "null"
   "sourcename", "The name of the source under analysis", "str", "yes", "null"
   "userName", "The name of the user performing the analysis", "str", "yes", "null" 
   "verboselvl", "| 0 ⇒ *CRITICAL* and *WARNING* messages are logged on the console.
   | 1 ⇒ *CRITICAL*, *WARNING* and *INFO* messages are logged on the console.
   | 2 ⇒ *CRITICAL*, *WARNING*, *INFO* and *DEBUG* messages are logged on the console",  "int", "no", 1


Section: *'selection'*
======================

The temporal, spatial and spectral binning of the data can be customized using the configuration options of this section.

The center of the *ROI* (region of interest) is defined by explicit Galactic sky coordinates (glon and glat).

.. csv-table::
   :header: "Option", "Description", "Type", "Default", "Required"
   :widths: 20, 100, 20, 20, 20

   "glat", "Center of the ROI ('*latitude*' or *'b'*)", "float", "null", "yes"
   "glon", "Center of the ROI ('*longitude*' or *'l'*)", "float", "null", "yes"
   "tmin", "Minimum time (in MJD or TT)", "float", "null", "yes"
   "tmax", "Maximum time (in MJD or TT)", "float", "null", "yes"
   "timetype", "| The date format of tmin and tmax.
   | Possibile values: [*'MJD'*, *'TT'*]", "str", "null", "yes"
   "timelist", "| it's a path to a file containing a list of time intervals in TT
   | format to generate maps
   | integrated within a time window.
   | If specified, *'tmin'* and *'tmax'* are ignored.", "str", "null", "no"
   "irf", "The IRF to be used for the analysis.", "str", "H0025", "no"
   "filtercode", "filtercode = 5 select G filtercode = 0 select G+L+S", "int", 5, "no"
   "fovradmin", "fovradmin < fovradmax", "int", 0, "no"
   "fovradmax", "fovradmax > fovradmin (dq = 0 is necessary for setting)", "int", 60, "no"
   "albedorad", "albedo selection cut (dq = 0 is necessary for setting)", "int", 80, "no"
   "dq", "| Data quality selection filter.
   | A combination of fovradmax and albedorad.
   | Possible values are [0,1,2,3,4,5,6,7,8,9]
   | dq = 0 -> albedorad and fovradmax are free and they must always be specified in setOption
   | dq = 1 -> albedorad=80, fovradmax=60
   | dq = 2 -> albedorad=80, fovradmax=50
   | dq = 3 -> albedorad=90, fovradmax=60
   | dq = 4 -> albedorad=90, fovradmax=50
   | dq = 5 -> albedorad=100, fovradmax=50
   | dq = 6 -> albedorad=90, fovradmax=40
   | dq = 7 -> albedorad=100, fovradmax=40
   | dq = 8 -> albedorad=90, fovradmax=30
   | dq = 9 -> albedorad=100, fovradmax=30", "int", 0, "no"
   "phasecode", "| Photon list selection parameter based
   | on the orbital phase. If 'None', the
   | automated selection is done following
   | the *'phasecode'* rule", "int", "null", "no"

Phasecode rule
^^^^^^^^^^^^^^

  - phasecode = 2 -> spinning mode, SAA excluded with AC counts method.
  - phasecode = 6 -> spinning mode, SAA excluded according to the magnetic field intensity (old definition of SAA, defined by TPZ)
  - phasecode = 18 -> pointing mode, SAA and recovery exluded.

It is suggested to use phasecode = 2 for data taken in spinning mode.

.. code-block:: ruby

    def setPhaseCode(tmax)
       if @phasecode == -1
          if tmax.to_f >= 182692800.0
             @phasecode = 6 #SPIN
          else
             @phasecode = 18 #POINTING
          end
       end
    end

filtercode rule
^^^^^^^^^^^^^^^^

A set of different on-board triggers enables the discrimination of background events (mainly cosmic rays in the AGILE Low Earth Orbit) from gamma-ray events. The data processing of the GRID events use an additional on-ground filters and provides a classification of each event:

- P : events classified as a charged particle and rejected
- G : events classified as gamma-ray photons. This is the most useful class for the analysis
- S : events classified as single-track: this is a special class of events with no separation between the electron and positron tracks
- L : limbo events, not clearly classified.

The events provided in the EVT files are of type G, S, and L. The AGILE team recommends to use the G class for scientific analysis. Only for gamma-ray bursts or other short transient events, and for pulsar timing analysis the G, S and L classes should be used together.


Section: *'maps'*
=================

These options control the behaviour of the sky maps generation tools.
The *'energybin'* and *'fovbinnumber'* options set the number of maps that are generated:

::

    number of maps = number of energy bins * fovbinnumber


The *'energybin'* option is a list of strings with the following format:

.. code-block:: yaml

    energybins:
      - 100, 1000
      - 1000, 3000

The *'fovbinnumber'* option sets the number of bins between *'fovradmin'* and *'fovradmax'* as:

::

    number of fov bins = (fovradmax-fovradmin)/fovbinnumber

.. note:: One map is generated for each possible combination between the *'energybin'* (emin, emax) and the *'fovbinnumber'* (fovmin, fovmax).
   The order of map generation is described by the following pseudocode:

   | For each fovmin..fovmax:
   |    For each emin..emax:
   |        generateMap(fovmin, fovmax, emin, emax)


.. csv-table::
   :header: "Option", "Description", "Type", "Default", "Required"
   :widths: 20, 100, 20, 20, 20

   "mapsize", "Width of the ROI in degrees","float", 40, "no"
   "useEDPmatrixforEXP", "Use the EDP matrix to generate the exposure map.", "boolean", "False", "no"
   "expstep", "| Step size of the exposure map, if 'None' it depends by
   | round(1 / binsize, 2) (e.g. 0.3->3, 0.25->4, 0.1->10)", "int", "None", "no"
   "spectralindex", "Spectral index of the exposure map", "float", 2.1, "no"
   "timestep", "LOG file step size of exposure map (LOG file are at 0.1s)", "float", 160, "no"
   "projtype", "Projection mode. Possible values: ['*WCS*']", "str", "WCS", "no"
   "proj", "| Spatial projection for WCS mode.
   | Possible values: ['*ARC*', '*AIT*']", "str", "ARC", "no"
   "skytype", "| gasmap:
   | 0) SKY000-1 + SKY000-5,
   | 1) gc_allsky maps + SKY000-5,
   | 2) SKY000-5
   | 3) SKY001 (old galcenter, binsize 0.1, full sky),
   | 4) SKY002 (new galcenter, binsize 0.1, full sky) ", "int", "4", "no"
   "binsize", "Spatial bin size in degrees", "float", 0.25, "no"
   "energybin", "The enegy bins of analysis. A list of value. 
   | To configure: 
   | 1) directly in the yaml configuration file; 
   | 2) Use the method e.g. ag.setOptions(energybins=[[100, 300], [500, 1000]]) 
   | 3) Use the method ag.setOptionEnergybin(value)",  "List<String>", "[100, 10000]", "no"
   "fovbinnumber", "| Number of bins between fovradmin and fovradmax.
   | Dim = (fovradmax-fovradmin)/fovbinnumber", "int", 1, "no"



Section: *'model'*
==================

The '*galcoeff*' and '*isocoeff*' options values can take the default value of null or they can be a a list of values separated by a comma.
If they are set to null it means they are free to change.

.. code-block:: yaml

    model:
      galcoeff: 0.8, 0.6, 0.5, 0.4
      isocoeff: 8, 10, 12, 14

In this case, you should pay attention on how the sky maps are generated: the
following example show which iso/gal coefficients are assigned to which map.

.. code-block:: yaml

    selection:
      fovradmin: 0
      fovradmax: 60

    maps:
      energybins:
        - 100, 300
        - 300, 1000
      fovbinnumber: 2

    model:
      galcoeff: 0.8, 0.6, 0.5, 0.4
      isocoeff: 8, 10, 12, 14

| **FOV bins:**
| (0, 30), (30, 60)


| **Map #1** has: fovmax:0  fovmax:30 emin:100 emax:300   galcoeff:0.8 isocoeff:8
| **Map #2** has: fovmax:0  fovmax:30 emin:300 emax:1000  galcoeff:0.6 isocoeff:10
| **Map #3** has: fovmax:30 fovmax:60 emin:100 emax:300   galcoeff:0.5 isocoeff:12
| **Map #4** has: fovmax:30 fovmax:60 emin:300 emax:1000  galcoeff:0.4 isocoeff:14



.. csv-table::
   :header: "Option", "Description", "Type", "Default", "Required"
   :widths: 20, 100, 20, 20, 20

   "modelfile", "| A file name that contains point
   | sources, diffuse and isotropic components", "string", "null", "yes"
   "galmode",  "int", 1, "no",
   "isomode", "int", 1, "no",
   "galcoeff", "set into .maplist if >= 0", "null, float or str", null, "no"
   "isocoeff", "set into .maplist if >= 0", "null, float or str", null, "no"

galcoeff and isocoeff
^^^^^^^^^^^^^^^^^^^^^^

galcoeff and isocoeff are the coefficients for the Galactic and isotropic diffuse emission components respectively. The values may be fixed during the fitting process or some or all of them may be optimized by allowing them to vary. Agilepy allows to evaluate these coefficient and fix them or to keep these coefficient free. 

Positive values are considered fixed, while negative values are free to vary starting from their absolute values. These coefficients are affected by the galmode and isomode coefficients described in the following section.

galmode and isomode
^^^^^^^^^^^^^^^^^^^

*'galmode'* and *'isomode'* are integer values describing how the corresponding
coefficients *'galcoeff'* or *'isocoeff'* found in all the lines of the maplist will be used:

| 0: all the coefficients are fixed.
| 1: all the coefficients are fixed if positive, variable if negative (the absolute value is the initial value). This is the default behaviour.
| 2: all the coefficients are variable, regardless of their sign.
| 3: all the coefficients are proportionally variable, that is the relative weight of their absolute value is kept.


Section: *'mle'*
================

The maximum likelihood estimation analysis is configured by the following options:

.. csv-table::
   :header: "Option", "Description", "Type", "Default", "Required"
   :widths: 20, 100, 20, 20, 20

   "ranal", "Radius of analysis", float, 10, No
   "ulcl", "Upper limit confidence level, expressed as sqrt(TS)", float, 2, No
   "loccl", "Source location contour confidence level (default 95 (%)confidence level). Possible values: [ *99*, *95*, *98*, *50*]", int, 95, No
   "fluxcorrection", "Correction of the flux taking into account the spectral model. Possible values: [0 (no correction), 1 (enable correction)]. ", int, 0, No

Exp-ratio evaluation options
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
See details in `this link <products.html#expratio>`_.

.. csv-table::
   :header: "Option", "Type", "Default", "Required", "Description"
   :widths: 20, 100, 20, 20, 20

   expratioevaluation, bool, yes, none, ""
   expratio_minthr, float, 0, none, ""
   expratio_maxthr, float, 15, none, ""
   expratio_size, float, 10, none, ""


Section: *'ap'*
===============

This section describes the configuration parameters for the Aperture Photometry analysis.

.. csv-table::
    :header: "Option", "Description", "Type", "Required", "Default"
    :widths: 20, 100, 20, 20, 20

    radius, "The radius of analysis", float, no, 3
    timeslot, "The size of the temporal bin", int, no, 3600


Section: *'plot'*
=================

This section defines the plotting configuration.

.. csv-table::
    :header: "Option", "Description", "Type", "Required", "Default"
    :widths: 20, 100, 20, 20, 20

    twocolumns, "The plot is adjusted to the size of a two column journal publication", boolean, False, no


CHANGELOG.md
# Changelog

## Release 1.6.5 (May 20, 2025)
* [#41](https://github.com/AGILESCIENCE/Agilepy/issues/41): Documentation update on installation procedures.
* [#48](https://github.com/AGILESCIENCE/Agilepy/issues/48): Fix a bug in the `entrypoint.sh` script for `agilepy` container that caused it to crash.
* [#46](https://github.com/AGILESCIENCE/Agilepy/issues/46): Implement `agilepy/scripts/start_test_rest_api.sh` to test the REST API with data download from SSDC, and use it in CI pipelines.
* [#37](https://github.com/AGILESCIENCE/Agilepy/issues/37): Introduce a CI pipeline that builds the `agilepy` image, runs the container and performs unit tests.
* [#44](https://github.com/AGILESCIENCE/Agilepy/issues/44): Fix CI pipelines to use `agilepy-recipe:BUILD26` instead of `BUILD25b6-v3`. Fix log directories so that tests can be executed without needing the root user.
* [#38](https://github.com/AGILESCIENCE/Agilepy/issues/38): Introduce a "Deprecated Option" Error to be thrown when the user tries to use "emin" and "emax" parameters for the AGAnalysis class. 
* Bug fix: the code to download data with the REST API used to crash due to a misplaced logging instruction.

## Release 1.6.4 (Oct 18, 2023, d3c067e)
* The old Jupyter notebooks have been replaced with JupyterLab (notebook v7)
* The Docker base image has been refactored. The user inside the container can be set to the user on the host to avoid file permission issues. The Root's python bindings are fixed. The unnecessary virtual environment has been removed. Ruby has been installed. The docker recipies have been moved inside the Agilepy repository. The Science Tools BUILD25b7 have been installed.
* The documentation has been improved.
* The Agilepy's python dependencies are now locked. 
* The "emin" and "emax" configuration parameters under "selection" have been deprecated. They are automatically calculated from the "energybins" configuration parameter.
* The "irf" configuration parameter has been introduced. It is used to select the IRF to be used in the analysis. The default value is "default".
* The logger has been refactored and improved. The duplicate logs issue has been fixed. The logging messages have been revisited.
* The test suite has been ported to pytest.
* The AGRatemeters tool has been added.
* The "advanced" API packaged has been introduced: it contains analysis scripts that are built over the "basic" API.

## Release 1.6.3
* Science Tools BUILD25b6, with changes on scripts to make them compatible with Python 3

## Release 1.6.2
* display off-axis comparison refined for GRB analysis
* minor fixes for aperture photometry

## Release 1.6.1
* Minor fixes
* root 6.26
* compilation with c++17

## Release 1.6.0
* (\#296) Now calcbkg updates the results if called multiple times
* (\#331) Docs updated
* (\#334) Minor fixes in print source
* (\#335) New parameter position in lightcurveMLE, default ellipse
* (\#336) Bugfix when computing fixflag   
* (\#342) Now it is possible to select [tmin - pastimewindow, tmax - tmin] as parameter om calcbkg
* (\#346) Minor fix in plots
* (\#347) Handling -nan parameter in multi tool
New anaconda recipes and docker container, environment and dependencies updated

## Release 1.5.1 (March 3, 2022)
* Minor bugfixes and doc updated

## Release 1.5.0 (February 24, 2022)
* (\#287) New Feature: automatic download of the AGILE public data using a REST API directly from SSDC!
* (\#234) Internal refactoring of Source class, new set/get interface for Source class. Docs updated.
* (\#291) Codacy coverage fixed
* (\#282) Dependencies updated and CI improved
* (\#287b) Brand new API for date conversion using Astropy Time
* (\#303) AG_spotfinder tool added
* Various bugfixes and improvements

## Release 1.4.2 (May 26, 2021)
* Minor hotfixes and docs updated

## Release 1.4.1 (April 26, 2021)
* minor bugfixes and docs updated

##  Release 1.4.0 (April 22, 2021)
* (\#234) Added new installation method with docker containers
* (\#267) Added several columns into lightcurvedata txt
* (\#272) Updated method for displaying columns of lightcurve data
* Various bugfixes and improvements

##  Release 1.3.0 (March 29, 2021)
* (\#133) New normalization functions in displaymaps
* (\#196) Notebooks have their own documentation section
* (\#201) The confFilePath parameter of the getConfiguration(..) method now supports environment variables.
* (\#202) A new dataset (for testing/validation purposes) has been included.
* (\#203) The "INDEX file with one line crash" bug has been fixed.
* (\#208) useEDPmatrixforEXP default value is now: False.
* (\#215) After calling mle(), each free parameter will be updated (if they need to).
* (\#217) The quickstart guide and the tutorial notebooks show how to change the value of a spectrum parameter of a source
* (\#231) New utility for filtering ap in Astroutils
* (\#238) New methods setOptionTimeMJD setOptionEnergybin
* (\#242) tdms progress bars 
* (\#252) New methods for setting dq
* (\#260) New lightcurvedata file
* (\#264) New plotting method for generic lightcurvedata columns
* (\#265) New parameter for plotting FERMI's lightcurve
* (anaconda package update) The installation time of Agilepy has been reduced a lot.

## Release 1.2.0 (December 10, 2020)
* (\#154) The method displayCtsSkyMaps() won't throw an error anymore in case of two maps to be drawn with singleMode=True.
* (\#152) The setOptions() call to increment the number of energy bins while galcoeff and isocoeff are equal = null  won't throw an error anymore.
* (\#161, \#165, \#166) Documentation updated.
* (\#160) Interactive plots in notebooks
* (\#170) Added APDisplayAGILEFermiComparison class to compare ap and offaxis results
* (\#172) Added Offaxis tool to compute agile and fermi offaxis
* (\#173) Refactoring: internal refactoring, AGEngVisibility1 renamed in AGEngAgileOffaxisVisibility, AGEngVisibility2 renamed in AGEngAgileFermiOffaxisVisibilityComparison, notebooks/documentation updated.  

agilepy/core/Parameters.py
# DESCRIPTION
#       Agilepy software
#
# NOTICE
#      Any information contained in this software
#      is property of the AGILE TEAM and is strictly
#      private and confidential.
#      Copyright (C) 2005-2020 AGILE Team.
#          Baroncelli Leonardo <leonardo.baroncelli@inaf.it>
#          Addis Antonio <antonio.addis@inaf.it>
#          Bulgarelli Andrea <andrea.bulgarelli@inaf.it>
#          Parmiggiani Nicolò <nicolo.parmiggiani@inaf.it>
#      All rights reserved.

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.


import os
from pathlib import Path

class Parameters:

    _datapath = Path(os.environ["AGILE"]).joinpath("model/scientific_analysis/data")

    if _datapath.exists():
        _calibrationFiles = os.listdir(_datapath)
    else:
        _calibrationFiles = []
        
    _skymapTemplate = "{}_{}.SKY002.{}_{}.disp.conv.sky.gz"

    _sarmatrixTemplate = "AG_GRID_G0017_{}_{}.sar.gz"
    _edpmatrixTemplate = "AG_GRID_G0017_{}_{}.edp.gz"
    _psdmatrixTemplate = "AG_GRID_G0017_{}_{}.psd.gz"

    _mapNameTemplate = "TN{}_TX{}_EN{}_EX{}_ID{}"

    _supported_irfs = ["I0025", "H0025"]

    _supportedEnergyBins = [[10000,50000],
                            [1000,3000],
                            [1000,50000],
                            [100,1000],
                            [100,10000],
                            [100,200],
                            [100,300],
                            [100,400],
                            [100,50000],
                            [200,400],
                            [200,50000],
                            [3000,10000],
                            [3000,50000],
                            [300,1000],
                            [300,10000],
                            [300,3000],
                            [300,50000],
                            [30,100],
                            [30,1000],
                            [30,400],
                            [30,50],
                            [30,50000],
                            [400,1000],
                            [400,10000],
                            [400,3000],
                            [400,50000],
                            [50,100],
                            [50,400],
                            [1000,10000],
                            [100,3000],
                            [50,1000],
                            [50,10000],
                            [50,300],
                            [50,3000],
                            [50,50000]                            
                          ]


    @staticmethod
    def getFilterCodeMapping(filtercode):
        """
        filtercode = 0 ALL --> T
        filtercode = 1 != L --> GS
        filtercode = 2 != G --> LS
        filtercode = 3 != G != L --> S (+P)
        filtercode = 4 != S --> GL
        filtercode = 5 != S != L = G (+P)
        filtercode = 6 != S != G = L (+P)
        filtercode = 7 != S != G <> L = null --> P (+P)                
        """
        if filtercode == 0:
            return "SFMT"
        elif filtercode == 1:
            return "SFMGS"
        elif filtercode == 2:
            return "SFMLS"
        elif filtercode == 3:
            return "SFMS"
        elif filtercode == 4:
            return "SFMGL"
        elif filtercode == 5:
            return "SFMG"
        elif filtercode == 6:
            return "SFML"
        elif filtercode == 7:
            return "SFMP"
        else:
            raise ValueError(f"Filter code {filtercode} is not supported")

    @staticmethod
    def getCalibrationMatrices(filtercode, irf):

        if irf not in Parameters._supported_irfs:
            raise ValueError(f"IRF {irf} is not supported")
        
        sarmatrix = Parameters._sarmatrixTemplate.format(Parameters.getFilterCodeMapping(filtercode), irf)
        edpmatrix = Parameters._edpmatrixTemplate.format(Parameters.getFilterCodeMapping(filtercode), irf)
        psdmatrix = Parameters._psdmatrixTemplate.format(Parameters.getFilterCodeMapping(filtercode), irf)

        if sarmatrix not in Parameters._calibrationFiles:
            raise ValueError(f"Calibration matrix {sarmatrix} is not available")
        if edpmatrix not in Parameters._calibrationFiles:
            raise ValueError(f"Calibration matrix {edpmatrix} is not available")
        if psdmatrix not in Parameters._calibrationFiles:
            raise ValueError(f"Calibration matrix {psdmatrix} is not available")

        return [Parameters._datapath.joinpath(sarmatrix), Parameters._datapath.joinpath(edpmatrix), Parameters._datapath.joinpath(psdmatrix)]


    @staticmethod
    def getSkyMap(emin, emax, filtercode, irf):
        if irf not in Parameters._supported_irfs:
            raise ValueError(f"IRF {irf} is not supported")        
        skymap = Parameters._skymapTemplate.format(emin, emax, Parameters.getFilterCodeMapping(filtercode), irf)
        if skymap in Parameters._calibrationFiles:
            skymapPath = Parameters._datapath.joinpath(skymap)
            if skymapPath.exists():
                return skymapPath
            else:
                raise ValueError(f"Sky map {skymap} is not available in {Parameters._datapath}")
        else:
            raise ValueError(f"Sky map {skymap} is not available in {Parameters._datapath}")
        
    @staticmethod
    def getCatEminEmax(catalogName):
        if catalogName == "2AGL":
            return (100, 10000)
        else:
            raise ValueError(f"Catalog {catalogName} is not supported")

    @staticmethod
    def checkEnergyBin(energyBin):
        energyBin = [int(energy) for energy in energyBin]
        if energyBin in Parameters._supportedEnergyBins:
            return True
        return False


    @staticmethod
    def getMapNamePrefix(tmin, tmax, emin, emax, glon, glat, stepi):
        return Parameters._mapNameTemplate.format(str(tmin),str(tmax),str(emin).zfill(5), str(emax).zfill(5), str(stepi).zfill(2))

    @staticmethod
    def getSupportedIRFs():
        return Parameters._supported_irfs

    @staticmethod
    def getSupportedEnergyBins():
        return Parameters._supportedEnergyBins
        
    @staticmethod
    def getSupportedCalibrationFiles():
        return Parameters._calibrationFiles